### PR TITLE
Adding new components stream and updating / improving circle

### PR DESCRIPTION
**Pull request by asaf-erlich**
# Description of change
Added components stream. This also involved improving circle and fixing the tests so that components are created if not enough exists. To do this I copied over the client code into a test_client file. I also created backlog tickets for the other paginated apis we aren't testing properly.

https://stitchdata.atlassian.net/browse/SRCE-4938

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Comment by kspeer825** [comment]
Test changes look good

File: tests/test_client.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        return self.session.send(request.prepare())
+
+    @backoff.on_exception(backoff.constant,
+                          RateLimitException,
+                          max_tries=10,
+                          interval=60)
+    def request(self, tap_stream_id, *args, **kwargs):
+        wait = (self.next_request_at - datetime.now()).total_seconds()
+        if wait > 0:
+            time.sleep(wait)
```
Comment (kspeer825) [inline tests/test_client.py:121]: Is this sleep necessary?

Comment (asaf-erlich) [reply tests/test_client.py:121]: So this was copied from the http.py file that is in the tap. It looks like it was added in 2017 by Buck Ryan, someone who is no longer on the team or it was open source? Anyway, I saw this at the top if you look at the constant used that creates the wait time:
```
# The project plan for this tap specified:
# > our past experience has shown that issuing queries no more than once every
# > 10ms can help avoid performance issues
TIME_BETWEEN_REQUESTS = timedelta(microseconds=10e3)
```
In my personal opinion this is why backoff is supposed to add incremental backoff with jitter, not add wait time like this... I mean I can test without it but I'd rather not spend a lot of time fiddling with something that works either. @dmosorast @cosimon @luandy64 thoughts?

Comment (dmosorast) [reply tests/test_client.py:121]: Discussed in Slack, but very likely not for this application. It was a gut feeling spacer from an earlier rendition of this integration.

File: tests/test_client.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    # Cache means it saves the result of this call forever
+    # It assumes project data does not change
+    @functools.lru_cache
+    def get_test_project(self):
+        return TestProjects(self._client).get_first()
+
+    # Cache means it saves the result of this call forever
+    # It assumes project data does not change
+    @functools.lru_cache
+    def get_test_user(self):
```
Comment (dmosorast) [inline tests/test_client.py:294]: Could this be defined either inside the function above or as a `__` method? To discourage outside use where the cache becomes an issue?

File: tests/test_config.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import os
```
Comment (dmosorast) [inline tests/test_config.py:1]: Whoah, this is cool.

**Review by dmosorast — APPROVED** [review]
Some outside-of-review comments, not necessary, seems good to me :shipit:

File: tests/test_client.py (hunk 2 above)
Comment (asaf-erlich) [reply tests/test_client.py:294]: Definitely agree with making the methods private. There's a reason I put it inside a method like this (intended only to be used by this instance) and not around the call to retrieve all the projects or users.

File: tests/test_config.py (hunk 3 above)
Comment (asaf-erlich) [reply tests/test_config.py:1]: Credit goes 100% to @kspeer825 - I copied this file almost 100% from tap-square where he wrote this test.

File: tests/test_client.py (hunk 2 above)
Comment (asaf-erlich) [reply tests/test_client.py]: Done in latest commit

File: tests/test_client.py (hunk 1 above)
Comment (asaf-erlich) [reply tests/test_client.py]: Removed in the latest commit. Again I only changed the test_client. I didn't remove this from the tap itself, even if I suspect we might want to do that I think this PR is big enough.

**Comment by asaf-erlich** [comment]
Ran one more time with api limit 70 to force the testing of code which created more component records. This resulted in the tap stats spitting this out (71 components):
```
        projects = 7    pks: {'id'}
        versions = 3    pks: {'id'}
        components = 71 pks: {'id'}
        project_types = 3       pks: {'key'}
        project_categories = 1  pks: {'id'}
        resolutions = 4 pks: {'id'}
        roles = 2       pks: {'id'}
        users = 20      pks: {'accountId'}
        issue_transitions = 95  pks: {'id'}
        issue_comments = 12     pks: {'id'}
        changelogs = 86 pks: {'id'}
        issues = 46     pks: {'id'}
        worklogs = 37   pks: {'id'}
```

Going to merge this yay

**Pull request merged**